Make --allow-polyfill-parser work even if installed php-ast version is outdated

Warn about it being outdated, and restart phan with that extension (and other problematic extensions) disabled.

I don't feel like it's worth the effort to try to make phan work perfectly with php-ast 0.0.1, for example.

See https://github.com/vimeo/psalm/blob/3.4.9/src/psalm.php#L288-L307 for how this could be done

Related to #2947

Disabling php-ast in the restart handle causes its own issues when opcache.enable_cli and opcache.file_cache are used, in php 7.3 - .

I'm assuming the real values before the restart conflict with the new php configs for some arrays/switch statements, not sure which.

Emptying out opcache.file_cache fixes those issues

```
AssertionError: Must be in class scope to get class in /path/to/programming/phan/src/Phan/Language/Context.php:558
Stack trace:
#0 /path/to/programming/phan/src/Phan/Parse/ParseVisitor.php(1444): Phan\Language\Context->getClassInScope(Object(Phan\CodeBase))
#1 /path/to/programming/phan/src/Phan/Parse/ParseVisitor.php(319): Phan\Parse\ParseVisitor->getContextClass()
#2 /path/to/programming/phan/src/Phan/Analysis.php(203): Phan\Parse\ParseVisitor->visitMethod(Object(ast\Node))
#3 /path/to/programming/phan/src/Phan/Analysis.php(230): Phan\Analysis::parseNodeInContext(Object(Phan\CodeBase), Object(Phan\Language\Context), Object(ast\Node))
#4 /path/to/programming/phan/src/Phan/Analysis.php(230): Phan\Analysis::parseNodeInContext(Object(Phan\CodeBase), Object(Phan\Language\Context), Object(ast\Node))
#5 /path/to/programming/phan/src/Phan/Analysis.php(230): Phan\Analysis::parseNodeInContext(Object(Phan\CodeBase), Object(Phan\Language\Context), Object(ast\Node))
#6 /path/to/programming/phan/src/Phan/Analysis.php(156): Phan\Analysis::parseNodeInContext(Object(Phan\CodeBase), Object(Phan\Language\Context), Object(ast\Node))
#7 /path/to/programming/phan/src/Phan/Phan.php(693): Phan\Analysis::parseFile(Object(Phan\CodeBase), 'internal', false, NULL, true)
#8 /path/to/programming/phan/src/Phan/Phan.php(136): Phan\Phan::loadConfiguredPHPExtensionStubs(Object(Phan\CodeBase))
#9 /path/to/programming/phan/src/phan.php(35): Phan\Phan::analyzeFileList(Object(Phan\CodeBase), Object(Closure))
#10 /path/to/programming/phan/phan(2): require_once('/path/to/pro...')
```

--- a/NEWS.md
+++ b/NEWS.md
@@ -18,6 +18,9 @@ New features(Analysis):
 Plugins:
 + Properly warn about redundant `@return` annotations followed by other annotation lines in `PHPDocRedundantPlugin`.
 
+Bug fixes:
++ Treat `Foo::class` as a reference to the class/interface/trait `Foo` (#2945)
+
 Jul 17 2019, Phan 2.2.6
 -----------------------
 

--- a/src/Phan/Analysis/PostOrderAnalysisVisitor.php
+++ b/src/Phan/Analysis/PostOrderAnalysisVisitor.php
@@ -1169,11 +1169,13 @@ class PostOrderAnalysisVisitor extends AnalysisVisitor
     public function visitClassName(Node $node) : Context
     {
         try {
-            (new ContextNode(
+            foreach ((new ContextNode(
                 $this->code_base,
                 $this->context,
                 $node->children['class']
-            ))->getClassList(false, ContextNode::CLASS_LIST_ACCEPT_OBJECT_OR_CLASS_NAME);
+            ))->getClassList(false, ContextNode::CLASS_LIST_ACCEPT_OBJECT_OR_CLASS_NAME) as $class) {
+                $class->addReference($this->context);
+            }
         } catch (CodeBaseException $exception) {
             $exception_fqsen = $exception->getFQSEN();
             $this->emitIssueWithSuggestion(

--- a/tests/plugin_test/expected/134_class_reference.php.expected
+++ b/tests/plugin_test/expected/134_class_reference.php.expected
@@ -1,0 +1,1 @@
+src/134_class_reference.php:4 PhanUnreferencedClass Possibly zero references to class \X\Y\Z\Other

--- a/tests/plugin_test/src/134_class_reference.php
+++ b/tests/plugin_test/src/134_class_reference.php
@@ -1,0 +1,10 @@
+<?php
+namespace X\Y\Z;
+class Foo{}
+class Other{}
+
+namespace A\B\C;
+use X\Y\Z\Foo;
+// Unconditionally treat Foo::class (ast\AST_CLASS_NAME) as a reference to that class,
+// in the same way using a constant from that class (ast\AST_CLASS_CONST) would count as a reference.
+$something = Foo::class;

